filebeat/tests/integration/filestream_delete_test.go

```diff
@@ -0,0 +1,677 @@
+	t *testing.T,
+	filebeat *integration.BeatProc,
+	es *api.APIHandler,
+	msgLogFilePath string) {
+
+	t.Run("can detect events not published", func(t *testing.T) {
+		// Wait a few times for the 'not finished' logs
+		notFinishedMsg := fmt.Sprintf(
+			"not all events from '%s' have been published, "+
+				"closing harvester",
+			msgLogFilePath)
+		for i := range 2 {
+			filebeat.WaitForLogs(
+				notFinishedMsg,
+				10*time.Second,
+				"[%d] Filebeat did not wait for the resource to be finished",
+				i,
+			)
+		}
+	})
+
+	// Reset odds in Mock-ES
+	if err := es.UpdateOdds(0, 0, 0, 0); err != nil {
+		t.Fatalf("cannot update mock-es odds: %s", err)
+	}
+}
+
+// testGracePeriod asserts:
+// - Grace period is interrupted if data is added to the file
+// - Grace period is respected if the file does not change
+func testGracePeriod(
+	t *testing.T,
+	filebeat *integration.BeatProc,
+	gracePeriod time.Duration,
+	dataAdded bool,
+	msgLogFilePath string) {
+
+	if dataAdded {
+		t.Run("grace period is interrupted when file changes", func(t *testing.T) {
+			gracePeriodMsg := fmt.Sprintf(
+				"all events from '%s' have been published, waiting for %s grace period",
+				msgLogFilePath, gracePeriod)
+
+			// This wait always takes 6.1s, I'm not quite sure why, probably it
+			// is caused by some backoff logic. Setting the backoff.* in the
+			// output is not enough. So we wait at least 10s here
+			filebeat.WaitForLogs(
+				gracePeriodMsg,
+				10*time.Second,
+				"did not start waiting for grace period")
+
+			// Wait 1/2 of the grace period, then add data to the file
+			time.Sleep(gracePeriod / 2)
+			integration.GenerateLogFile(t, msgLogFilePath, 5, true)
+
+			// Wait for the message of file size changed
+			changedMsg := fmt.Sprintf("'%s' was updated, won't remove. Closing harvester", msgLogFilePath)
+			filebeat.WaitForLogs(changedMsg, time.Second, "filestream did detect the file change")
+
+			// Make sure the harvester is closed
+			filebeat.WaitForLogs("Stopped harvester for file", time.Second, "harvester was not closed")
+			filebeat.WaitForLogs("Closing reader of filestream", time.Second, "reader was not closed")
+		})
+	}
+
+	t.Run("grace period is respected", func(t *testing.T) {
+		msg := fmt.Sprintf("'%s' removed", msgLogFilePath)
+		filebeat.WaitForLogs(msg, 30*time.Second, "file removed log entry not found")
+		removedMsg := filebeat.GetLastLogLine(msg)
+
+		gracePeriodMsg := fmt.Sprintf(
+			"all events from '%s' have been published, waiting for %s grace period",
+			msgLogFilePath, gracePeriod)
+		beforeWait := filebeat.GetLastLogLine(gracePeriodMsg)
+
+		delta := timeBetweenLogEntries(t, beforeWait, removedMsg)
+		if delta < gracePeriod {
+			t.Errorf("grace period of %s was not respected", gracePeriod)
+			t.Log("grace period waiting calculated based on the following log entries:")
+			t.Log("First :", beforeWait)
+			t.Log("Second:", removedMsg)
+		}
+	})
+}
+
+func TestFilestreamDeleteRestart(t *testing.T) {
```

Comment:
It's still not clear to me what additional test coverage we get from the restart in this test case. There is value in testing Filebeat starting up with the file in various starting states (already EOF and acknowledged for example) but the deletion logic isn't affected by restarts at all as far as I can tell.

The reason I'm asking is that this is an integration test, and integration tests are generally more expensive for us to maintain long term, so if we can get rid of one that's valuable.

Reply:
Lee requested it when he was reviewing, see the [request](https://github.com/elastic/beats/pull/43368#issuecomment-2825306933) and the following 3 comments. Even though it is an integration test it is a simple one, so I'm not worried about maintaining it.

It is an extra test case covering the shutdown and restart of the input and all of its components that need to interact to make this feature work.

If you really insist, I can remove it. But my vote is to keep it.

Reply:
Reading this comment, it seems like this should be a test that a file that was previously ingest completely is deleted after the grace period. The restart itself isn't testing, it's testing it can resume based on registry state but testing that doesn't require a restart in the test, just that the registry file exists with the required conditions. You are just stopping filebeat in the middle of ingesting data to pre-populate the registry, if anything the test should have a more specific name. We aren't trying to stress test Filebeat here.

The part of Filebeat that is sensitive to restarts is the registry file and that is functionally separate from this.